### PR TITLE
Lesson 13: Magento Cookies, Session, Session storage

### PR DESCRIPTION
**Pull request by renegade-d3v**
1. Implement Redis for all environments
2. Add formkey validation
3. Save Customer ID to the database when the request is sent
4. Create ProducID request controller, modify request Model
5. Modify Block, Templates, JS, i18n files for showing "Alredy requested!" message in the tab on the Product Page.
6. Remove Unique Index on email + webstie_id from schema files
7. Make up DB dump

### CODE REVIEW

File: app/code/AndriiShkrebtii/RegularCustomer/i18n/ru_RU.csv [hunk 1]
```diff
@@ -16,3 +16,4 @@ declined,отклонен
 "Request status:","Статус запроса:"
 "No customer requests","Нет клиентских запросов"
 "Regular Customer Page","Страница постоянного клиента"
+"Already requested!","Скидка уже запрашена!"
```
Comment (maksymz) [inline app/code/AndriiShkrebtii/RegularCustomer/i18n/ru_RU.csv]: @jimmywilson111 `запрашена`? Better to translate this as `Запит вже надіслано!`

File: app/code/AndriiShkrebtii/RegularCustomer/view/frontend/web/js/message.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+         */
+        _create: function () {
+            $(document).on('andrii_shkrebrii_regular_customer_show_message', this.showMessage.bind(this));
+        },
+
+        /**
+         * Generate event to show message
+         */
+        showMessage: function () {
+            $(this.element).css('display', 'inline-block');
```
Comment (maksymz) [inline app/code/AndriiShkrebtii/RegularCustomer/view/frontend/web/js/message.js]: @jimmywilson111 Do not decide in JS  which type of display to use. This is for CSS. Should add `display:none;` or remove it, but CSS rules must be written in the LESS files/. In JS you should better operate visibility and nothing else.

File: app/code/AndriiShkrebtii/RegularCustomer/Setup/Patch/Data/RemoveUniqueIndex.php__ [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        $this->schemaSetup = $schemaSetup;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply()
+    {
+        $this->schemaSetup->getConnection()->startSetup();
+        //
```
Comment (maksymz) [inline app/code/AndriiShkrebtii/RegularCustomer/Setup/Patch/Data/RemoveUniqueIndex.php__]: @jimmywilson111 Implementation missed

**Review by maksymz — CHANGES_REQUESTED** [review]

File: app/code/AndriiShkrebtii/RegularCustomer/etc/frontend/di.xml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
+    <type name="AndriiShkrebtii\RegularCustomer\Controller\Index\Request">
```
[the remark this answers is not in the document]
Comment (renegade-d3v) [reply app/code/AndriiShkrebtii/RegularCustomer/etc/frontend/di.xml:8]: Done ✔

File: app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/ProductID.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+namespace AndriiShkrebtii\RegularCustomer\Controller\Index;
+
+use Magento\Framework\Controller\Result\Json as JsonResponse;
+
+class ProductID implements \Magento\Framework\App\Action\HttpGetActionInterface
+{
+    /**
+     * @var \Magento\Framework\Controller\Result\JsonFactory $jsonResponseFactory
+     */
+    private $jsonResponseFactory;
```
[the remark this answers is not in the document]
Comment (renegade-d3v) [reply app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/ProductID.php:14]: Done ✔

File: app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/Request.php [hunk 6]
```diff
@@ -74,21 +91,29 @@ public function execute(): JsonResponse
         // @TODO: add form key validation and hideIt validation
         // @TODO: add Google Recaptcha to the form
         $response = $this->jsonResponseFactory->create();
-
+        $formSaved = false;
         try {
```
[the remark this answers is not in the document]
Comment (renegade-d3v) [reply app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/Request.php:96]: Done ✔

File: app/code/AndriiShkrebtii/RegularCustomer/i18n/ru_RU.csv (hunk 1 above)
Comment (renegade-d3v) [reply app/code/AndriiShkrebtii/RegularCustomer/i18n/ru_RU.csv]: Done ✔

File: app/code/AndriiShkrebtii/RegularCustomer/view/frontend/web/js/message.js (hunk 2 above)
Comment (renegade-d3v) [reply app/code/AndriiShkrebtii/RegularCustomer/view/frontend/web/js/message.js]: Done ✔

File: app/code/AndriiShkrebtii/RegularCustomer/Setup/Patch/Data/RemoveUniqueIndex.php__ (hunk 3 above)
Comment (renegade-d3v) [reply app/code/AndriiShkrebtii/RegularCustomer/Setup/Patch/Data/RemoveUniqueIndex.php__]: As soon as possible there will be a patch to fix the removal of the constraint from database.

File: app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/Request.php [hunk 7]
```diff
@@ -74,22 +91,35 @@ public function execute(): JsonResponse
...
                 ->setEmail($this->request->getParam('email'))
-                ->setWebsiteId($this->storeManager->getStore()->getWebsiteId())
+                ->setCustomerId((int)$this->customerSession->getCustomerId())
+                ->setProductId($this->request->getParam('productId'))
+                ->setWebsiteId((int)$this->storeManager->getStore()->getWebsiteId())
                 ->setStatus(DiscountRequest::STATUS_PENDING);
             $this->discountRequestResource->save($discountRequest);
-            $message = __('You request for registration in program was accepted!');
+            $formSaved = true;
+
```
Comment (maksymz) [inline app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/Request.php]: @jimmywilson111 Extra empty line here

File: app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/Request.php [hunk 8]
```diff
@@ -74,22 +91,35 @@ public function execute(): JsonResponse
...
+        } catch (\InvalidArgumentException $e) {
```
Comment (maksymz) [inline app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/Request.php]: @jimmywilson111 Extra empty line here

File: app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/Request.php [hunk 9]
```diff
@@ -74,22 +91,35 @@ public function execute(): JsonResponse
...
 
         } catch (\Exception $e) {
+
```
Comment (maksymz) [inline app/code/AndriiShkrebtii/RegularCustomer/Controller/Index/Request.php]: @jimmywilson111 Extra empty line here

File: app/code/AndriiShkrebtii/RegularCustomer/Setup/Patch/Data/RemoveUniqueIndex.php__ (hunk 3 above)
Comment (maksymz) [reply app/code/AndriiShkrebtii/RegularCustomer/Setup/Patch/Data/RemoveUniqueIndex.php__]: @jimmywilson111 Can't accept merge request without the complete patch

File: app/code/AndriiShkrebtii/RegularCustomer/etc/frontend/di.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (maksymz) [reply app/code/AndriiShkrebtii/RegularCustomer/etc/frontend/di.xml:8]: @jimmywilson111 ??? 
There are no changes and still no proxy for `customerSession` parameter in the block.

**Review by maksymz — CHANGES_REQUESTED** [review]

**Pull request merged**